File: server/src/media_storage_server/compression.py
"""
media-storage : compression
===========================

Offers efficient handlers for compressing and decompressing data, using
file-like objects (often tempfiles).

This module is shared by every Python facet of the media-storage project and
changes to one instance should be reflected in all.

Usage
-----

(De)compressors may be called explicitly or retrieved with one of the getter
functions.
 
Legal
-----

This file is part of the LGPLed subset of the media-storage project.
This module is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published
by the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU General Public License and
GNU Lesser General Public License along with this program. If not, see
<http://www.gnu.org/licenses/>.
 
(C) Neil Tallim, 2011
"""
import bz2
import logging
import tempfile
import zlib

try:
    import lzma
except ImportError:
    lzma = None
    
#Compression type constants
COMPRESS_NONE = None
COMPRESS_BZ2 = 'bz2'
COMPRESS_GZIP = 'gzip'
COMPRESS_LZMA = 'lzma'

SUPPORTED_FORMATS = [COMPRESS_GZIP, COMPRESS_BZ2]
if lzma:
    SUPPORTED_FORMATS.append(COMPRESS_LZMA)
SUPPORTED_FORMATS = tuple(SUPPORTED_FORMATS)

_MAX_SPOOLED_FILESIZE = 1024 * 256 #Allow up to 256k in memory
_BUFFER_SIZE = 1024 * 32 #Work with 32k chunks

_logger = logging.getLogger('media_server-compression')

def get_compressor(format):
    """
    Returns a callable that accepts a file-like object and returns a compressed version of the
    file's contents as a file-like object.
    
    `format` is the format to which conversion should occur, one of the compression type constants.
    """
    if format is COMPRESS_NONE:
        return (lambda x:x)
    elif format == COMPRESS_GZIP:
        return compress_gzip
    elif format == COMPRESS_BZ2:
        return compress_bz2
    elif format == COMPRESS_LZMA and lzma:
        return compress_lzma
    raise ValueError(format + " is unsupported")
    
def get_decompressor(format):
    """
    Returns a callable that accepts a file-like object and returns a decompressed version of the
    file's contents as a file-like object.
    
    `format` is the format from which conversion should occur, one of the compression type constants.
    """
    if format is COMPRESS_NONE:
        return (lambda x:x)
    elif format == COMPRESS_GZIP:
        return decompress_gzip
    elif format == COMPRESS_BZ2:
        return decompress_bz2
    elif format == COMPRESS_LZMA and lzma:
        return decompress_lzma
    raise ValueError(format + " is unsupported")
    
def _process(data, handler, flush_handler):
    """
    Iterates over the given `data`, reading a reasonable number of bytes, passing them through the
    given (de)compression `handler`, and writing the output to a temporary file, which is ultimately
    returned (seeked to 0).
    
    If an exception occurs, it is raised directly.
    """
    try:
        temp = tempfile.SpooledTemporaryFile(_MAX_SPOOLED_FILESIZE)
        while True:
            chunk = data.read(_BUFFER_SIZE)
            if chunk:
                chunk = handler(chunk)
                if chunk:
                    temp.write(chunk)
            else:
                if flush_handler:
                    chunk = flush_handler()
                    if chunk:
                        temp.write()
                break
        temp.flush()
        temp.seek(0)
        return temp
    except Exception as e:
        _logger.error("A problem occurred during (de)compression: %(error)s" % {
         'error': str(e),
        })
        raise
        
def compress_bz2(data):
    """
    Compresses the given file-like object `data` with the bz2 algorithm, returning a file-like
    object and its size in a tuple.
    
    Any exceptions are raised directly.
    """
    _logger.debug("Compressing data with bz2...")
    compressor = bz2.BZ2Compressor()
    return _process(data, compressor.compress, compressor.flush)

def decompress_bz2(data):
    """
    Decompresses the given file-like object `data` with the bz2 algorithm, returning a file-like
    object and its size in a tuple.
    
    Any exceptions are raised directly.
    """
    _logger.debug("Decompressing data with bz2...")
    decompressor = bz2.BZ2Decompressor()
    return _process(data, decompressor.decompress, None)
    
def compress_gzip(data):
    """
    Compresses the given file-like object `data` with the gzip algorithm, returning a file-like
    object and its size in a tuple.
    
    Any exceptions are raised directly.
    """
    _logger.debug("Compressing data with gzip...")
    compressor = zlib.compressobj()
    return _process(data, compressor.compress, compressor.flush)
    
def decompress_gzip(data):
    """
    Decompresses the given file-like object `data` with the gzip algorithm, returning a file-like
    object and its size in a tuple.
    
    Any exceptions are raised directly.
    """
    _logger.debug("Decompressing data with gzip...")
    decompressor = zlib.decompressobj()
    return _process(data, decompressor.decompress, decompressor.flush)
    
if lzma: #If the module is unavailable, don't even define the functions
    def compress_lzma(data):
        """
        Compresses the given file-like object `data` with the lzma algorithm, returning a file-like
        object and its size in a tuple.
        
        Any exceptions are raised directly.
        
        This function is not available if no LZMA library is present.
        """
        _logger.debug("Compressing data with lzma...")
        compressor = lzma.LZMACompressor()
        return _process(data, compressor.compress, compressor.flush)
        
    def decompress_lzma(data):
        """
        Decompresses the given file-like object `data` with the lzma algorithm, returning a
        file-like object and its size in a tuple.
        
        Any exceptions are raised directly.
        
        This function is not available if no LZMA library is present.
        """
        _logger.debug("Decompressing data with lzma...")
        decompressor = lzma.LZMADecompressor()
        return _process(data, decompressor.decompress, decompressor.flush)
        

File: clients/c-sharp/Libraries/Compression.cs
//  
//  Author:
//       Neil Tallim <flan@uguu.ca>
// 
//  Copyright (c) 2012 Ivrnet, inc.
// 
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
// 
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
// 
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
// 

namespace MediaStorage{
    /// <summary>
    /// Specifies the type of compression to apply to the data while it's in storage
    /// </summary>
    public enum COMPRESSION{
        /// <summary>
        /// No compression is applied
        /// </summary>
        NONE,
        /// <summary>
        /// Gzip is fast and does a decent job of compressing content like documents
        /// </summary>
        GZ,
        /// <summary>
        /// BZ2 offers a good mix of speed and compression of all content; it has no client-side CLR handling
        /// </summary>
        BZ2,
        /// <summary>
        /// LZMA is slow, but compresses very aggressively; it has no client-side CLR handling
        /// </summary>
        LZMA,
    }
}

namespace MediaStorage.Libraries{
    /// <summary>
    /// Exposes compression methods.
    /// </summary>
    internal static class Compression{
        /// <summary>
        /// An enumerable collection of all locally supported formats.
        /// </summary>
        internal static System.Collections.Generic.IList<COMPRESSION> SupportedFormats = new System.Collections.Generic.List<COMPRESSION>(new COMPRESSION[]{COMPRESSION.GZ});

        /// <summary>
        /// Resolves a compression format-enum, given a string.
        /// </summary>
        /// <param name='compression'>
        /// The compression-format to be identified; internally evaluated in lower-case.
        /// </param>
        /// <exception cref="System.ArgumentException">
        /// An unknown format was indicated.
        /// </exception>
        internal static COMPRESSION ResolveCompressionFormat(string compression){
            if(compression == null){
                return COMPRESSION.NONE;
            }
            switch(compression.ToLower()){
                case "gz":
                    return COMPRESSION.GZ;
                case "bz2":
                    return COMPRESSION.BZ2;
                case "lzma":
                    return COMPRESSION.LZMA;
                default:
                    throw new System.ArgumentException("Unsupported compression format: " + compression);
            }
        }

        /// <summary>
        /// Gets the appropriate compressor-function for the indicated format.
        /// </summary>
        /// <param name='compression'>
        /// The type of compressor to retrieve.
        /// </param>
        /// <exception cref="System.ArgumentException">
        /// An unsupported format was indicated.
        /// </exception>
        internal static System.Func<System.IO.Stream, System.IO.Stream> GetCompressor(COMPRESSION compression){
            if(compression == COMPRESSION.NONE){
                return Compression.NullHandler;
            }else if(compression == COMPRESSION.GZ){
                return Compression.CompressGz;
            }
            throw new System.ArgumentException(compression.ToString() + " is unsupported");
        }

        /// <summary>
        /// Gets the appropriate decompressor-function for the indicated format.
        /// </summary>
        /// <param name='compression'>
        /// The type of decompressor to retrieve.
        /// </param>
        /// <exception cref="System.ArgumentException">
        /// An unsupported format was indicated.
        /// </exception>
        internal static System.Func<System.IO.Stream, System.IO.Stream> GetDecompressor(COMPRESSION compression){
            if(compression == COMPRESSION.NONE){
                return Compression.NullHandler;
            }else if(compression == COMPRESSION.GZ){
                return Compression.DecompressGz;
            }
            throw new System.ArgumentException(compression.ToString() + " is unsupported");
        }

        /// <summary>
        /// Returns the data unmodified. (New <see cref="TempFileStream"/> for semantic equivalence)
        /// </summary>
        /// <param name='data'>
        /// The data to be (de)compressed.
        /// </param>
        private static System.IO.Stream NullHandler(System.IO.Stream data){
            System.IO.Stream tfs = new TempFileStream();

            data.CopyTo(tfs);
            tfs.Seek(0, System.IO.SeekOrigin.Begin);

            return tfs;
        }

        /// <summary>
        /// Returns the data compressed using Gzip.
        /// </summary>
        /// <param name='data'>
        /// The data to be compressed.
        /// </param>
        private static System.IO.Stream CompressGz(System.IO.Stream data){
            System.IO.Compression.GZipStream gz = new System.IO.Compression.GZipStream(data, System.IO.Compression.CompressionMode.Compress);
            TempFileStream tfs = new TempFileStream();

            gz.CopyTo(tfs);
            tfs.Seek(0, System.IO.SeekOrigin.Begin);

            return tfs;
        }

        /// <summary>
        /// Returns the data decompressed using Gzip.
        /// </summary>
        /// <param name='data'>
        /// The data to be decompressed.
        /// </param>
        private static System.IO.Stream DecompressGz(System.IO.Stream data){
            System.IO.Compression.GZipStream gz = new System.IO.Compression.GZipStream(data, System.IO.Compression.CompressionMode.Decompress);
            TempFileStream tfs = new TempFileStream();

            gz.CopyTo(tfs);
            tfs.Seek(0, System.IO.SeekOrigin.Begin);

            return tfs;
        }
    }
}
